core/src/test/scala/integration/kafka/api/ProducerFailureHandlingTest.scala

```diff
@@ -175,23 +176,23 @@ class ProducerFailureHandlingTest extends KafkaServerTestHarness {
   }
 
   /**
-   *  The send call with invalid partition id should throw KafkaException caused by IllegalArgumentException
-   */
+    * Send with invalid partition id should throw KafkaException when partition is higher than the
+    * upper bound of partitions and IllegalArgumentException when partition is negative
```

Comment:
I think I wasn't clear. Say you add partitions and then call send. We know the producer will do a metadata refresh if it is stale. But that by itself doesn't tell us that we will get the behaviour we want. The broker can return stale metadata as well, metadata propagation is asynchronous after all. The end to end test would capture the use case better. Is that clearer?